### PR TITLE
2013 drupal jen4

### PR DESCRIPTION
**Pull request by jenlampton**
PR for https://github.com/backdrop/backdrop-issues/issues/2013

### CODE REVIEW

File: core/modules/field/modules/number/number.module [hunk 1]
```diff
@@ -203,6 +203,8 @@ function number_field_formatter_settings_form($field, $instance, $view_mode, $fo
   $settings = $display['settings'];
   $element = array();
 
+  $element = array();
```
Comment (klonos) [inline core/modules/field/modules/number/number.module]: This is from https://www.drupal.org/node/1732906 / http://cgit.drupalcode.org/drupal/commit/?id=11dac6e but I don't think we need it (same line just 2 lines above). Must have fixed it at some earlier point.

File: core/modules/user/tests/user.test [hunk 2]
```diff
@@ -1867,6 +1880,78 @@ class UserEditTestCase extends BackdropWebTestCase {
...
+    $user1->pass_raw = '0';
+    $this->backdropLogin($user1);
+    $this->backdropLogout();
+  }
+}
+
+/**
+ * Tests editing a user account with and without a form rebuild.
+ */
+class UserEditRebuildTestCase extends DrupalWebTestCase {
```
Comment (klonos) [inline core/modules/user/tests/user.test]: `extends BackdropWebTestCase`

File: core/modules/user/user.module [hunk 3]
```diff
@@ -853,7 +853,7 @@ function user_validate_current_pass(&$form, &$form_state) {
     // that prevent them from being empty if they are changed.
     if ((strlen(trim($form_state['values'][$key])) > 0) && ($form_state['values'][$key] != $account->$key)) {
       require_once BACKDROP_ROOT . '/' . settings_get('password_inc', 'core/includes/password.inc');
-      $current_pass_failed = empty($form_state['values']['current_pass']) || !user_check_password($form_state['values']['current_pass'], $account);
+      $current_pass_failed = strlen(trim($form_state['values']['current_pass'])) == 0 || !user_check_password($form_state['values']['current_pass'], $account);
```
Comment (klonos) [inline core/modules/user/user.module]: In the respective commit in d.org, there was another single-line change in the `user_save()` function that you seem to be missing @jenlampton

**Review by klonos — CHANGES_REQUESTED** [review]

File: core/modules/user/user.module (hunk 3 above)
Comment (klonos) [reply core/modules/user/user.module]: I now see that this function has changed in Backdrop. Is that change irrelevant then?

```diff
 function user_save($account, $edit = array(), $category = 'account') {
   $transaction = db_transaction();
   try {
-    if (!empty($edit['pass'])) {
+    if (isset($edit['pass']) && strlen(trim($edit['pass'])) > 0) {
       // Allow alternate password hashing schemes.
       require_once DRUPAL_ROOT . '/' . variable_get('password_inc', 'includes/password.inc');
       $edit['pass'] = user_hash_password(trim($edit['pass']));
```

Comment (jenlampton) [reply core/modules/user/user.module]: I don't think "empty" is a good a check as "length = 0", this is still an improvement.

**Comment by jenlampton** [comment]
Rebased and pushed a new PR since test fail links are 404 on this issue too.

File: core/modules/user/tests/user.test [hunk 4]
```diff
@@ -1867,6 +1880,78 @@ class UserEditTestCase extends BackdropWebTestCase {
...
+ */
+class UserEditRebuildTestCase extends BackdropWebTestCase {
+
+  public static function getInfo() {
+    return array(
+      'name' => 'User edit with form rebuild',
+      'description' => 'Test user edit page when a form rebuild is triggered.',
+      'group' => 'User',
+    );
+  }
```
Comment (quicksketch) [inline core/modules/user/tests/user.test]: This getInfo() method should be converted to user.tests.info.

**Review by quicksketch — CHANGES_REQUESTED** [review]
There is a single legit test failure, as well as one test that needs to be converted to use Backdrop's .info method of defining tests.

**Review by quicksketch — APPROVED** [review]

**Comment by quicksketch** [comment]
Thanks for doing all these updates! There is still one test failure here though. Seems to be related to the password length changes. I'll see if I can correct that and push out an update.

**Comment by quicksketch** [comment]
I think this is good to go. Test failures seem to be unrelated, a side-effect from some other cross-ports. I've filed and issue at https://github.com/backdrop/backdrop-issues/issues/2939 to fix those failures, then we can retest here.

**Pull request merged**